htlcswitch/decayedlog.go

```diff
@@ -0,0 +1,406 @@
+
+	// Close boltdb.
+	d.db.Close()
+
+	return nil
+}
+
+// garbageCollector deletes entries from sharedHashBucket whose expiry height
+// has already past. This function MUST be run as a goroutine.
+func (d *DecayedLog) garbageCollector(epochClient *chainntnfs.BlockEpochEvent) {
+	defer d.wg.Done()
+	defer epochClient.Cancel()
+
+	for {
+		select {
+		case epoch, ok := <-epochClient.Epochs:
+			if !ok {
+				// Block epoch was canceled, shutting down.
+				log.Infof("Block epoch canceled, " +
+					"decaying hash log shutting down")
+				return
+			}
+
+			// Perform a bout of garbage collection using the
+			// epoch's block height.
+			height := uint32(epoch.Height)
+			numExpired, err := d.gcExpiredHashes(height)
+			if err != nil {
+				log.Errorf("unable to expire hashes at "+
+					"height=%d", height)
+			}
+
+			if numExpired > 0 {
+				log.Infof("Garbage collected %v shared "+
+					"secret hashes at height=%v",
+					numExpired, height)
+			}
+
+		case <-d.quit:
+			// Received shutdown request.
+			log.Infof("Decaying hash log received " +
+				"shutdown request")
+			return
+		}
+	}
+}
+
+// gcExpiredHashes purges the decaying log of all entries whose CLTV expires
+// below the provided height.
+func (d *DecayedLog) gcExpiredHashes(height uint32) (uint32, error) {
+	var numExpiredHashes uint32
+
+	err := d.db.Batch(func(tx *bolt.Tx) error {
+		numExpiredHashes = 0
+
+		// Grab the shared hash bucket
+		sharedHashes := tx.Bucket(sharedHashBucket)
+		if sharedHashes == nil {
+			return fmt.Errorf("sharedHashBucket " +
+				"is nil")
+		}
+
+		var expiredCltv [][]byte
+		if err := sharedHashes.ForEach(func(k, v []byte) error {
+			// Deserialize the CLTV value for this entry.
+			cltv := uint32(binary.BigEndian.Uint32(v))
+
+			if cltv < height {
+				// This CLTV is expired. We must add it to an
+				// array which we'll loop over and delete every
+				// hash contained from the db.
+				expiredCltv = append(expiredCltv, k)
+				numExpiredHashes++
+				fmt.Println("inc")
+			}
+
+			return nil
+		}); err != nil {
+			return err
+		}
+
+		// Delete every item in the array. This must
+		// be done explicitly outside of the ForEach
+		// function for safety reasons.
+		for _, hash := range expiredCltv {
+			err := sharedHashes.Delete(hash)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return 0, nil
+	}
+
+	return numExpiredHashes, nil
+}
+
+// Delete removes a <shared secret hash, CLTV> key-pair from the
+// sharedHashBucket.
+func (d *DecayedLog) Delete(hash *sphinx.HashPrefix) error {
+	return d.db.Batch(func(tx *bolt.Tx) error {
+		sharedHashes := tx.Bucket(sharedHashBucket)
+		if sharedHashes == nil {
+			return ErrDecayedLogCorrupted
+		}
+
+		return sharedHashes.Delete(hash[:])
+	})
+}
+
+// Get retrieves the CLTV of a processed HTLC given the first 20 bytes of the
+// Sha-256 hash of the shared secret.
+func (d *DecayedLog) Get(hash *sphinx.HashPrefix) (uint32, error) {
+	var value uint32
```

Comment:
Similar comment here as well w.r.t changing the default return value in the case of an error.